src/cosim/algorithm/ecco_algorithm.cpp

```diff
@@ -0,0 +1,600 @@
+/*
+ *  This Source Code Form is subject to the terms of the Mozilla Public
+ *  License, v. 2.0. If a copy of the MPL was not distributed with this
+ *  file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+#include "cosim/algorithm/ecco_algorithm.hpp"
+
+#include "cosim/error.hpp"
+#include "cosim/exception.hpp"
+#include "cosim/log/logger.hpp"
+#include "cosim/time.hpp"
+#include "cosim/utility/thread_pool.hpp"
+
+#include <algorithm>
+#include <cstdlib>
+#include <iostream>
+#include <numeric>
+#include <sstream>
+#include <unordered_map>
+#include <vector>
+
+
+namespace cosim
+{
+
+class ecco_algorithm::impl
```

Comment:
Not need to be addressed in this PR, but I think a lot of methods here are duplicates from `fixed_step_algorithm`. Not sure if it would cost more compilation time but we can create a base class for the algorithm `impl` for reusing those methods.

Reply:
There may be a potential to avoid duplication, but I don't think that inheritance is the way to exploit it, and I don't think we need to worry about it quite yet. There are two rules of thumb I like to follow in cases like this:

* Prefer [composition over inheritance](https://en.wikipedia.org/wiki/Composition_over_inheritance)
* The [rule of three](https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming))